File: onlinejudge/leetcode/2226.java
class Solution {
    public int maximumCandies(int[] candies, long k) {
        long total = sum(candies);
        if (total < k) {
            return 0;
        }

        int left = 1;
        int right = Arrays.stream(candies).max().getAsInt();

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (canDistribute(candies, k, mid)) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return right;
    }

    private boolean canDistribute(int[] candies, long k, int pileSize) {
        long count = 0;
        for (int candy : candies) {
            count += candy / pileSize;
            if (count >= k) {
                return true;
            }
        }
        return false;
    }

    private long sum(int[] arr) {
        long total = 0;
        for (int num : arr) {
            total += num;
        }
        return total;
    }
}

File: onlinejudge/leetcode/2226.rb
# @param {Integer[]} candies
# @param {Integer} k
# @return {Integer}
def maximum_candies(candies, k)
  # validate
  return 0 if candies.sum < k

  left = 1
  right = candies.max

  while left <= right
    mid = left + (right - left) / 2
    if can_distribute(candies, k, mid)
      left = mid + 1
    else
      right = mid - 1
    end
  end

  right
end

private

def can_distribute(candies, k, pile_size)
  count = candies.sum { |candy| candy / pile_size }
  count >= k
end